File: javacode/src/main/java/com/xiong/LeetCode/ArrayProblems/Leet18_fourSum.java
package com.xiong.LeetCode.ArrayProblems;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @author ：xiongcong
 * @date ：Created in 2020/10/5 10:29
 * @description：
 * @modified By：
 * @version: $
 */
public class Leet18_fourSum {

    public List<List<Integer>> fourSum(int[] nums, int target) {

        List<List<Integer>> ans = new ArrayList<>();
        Arrays.sort(nums);

        int n = nums.length;
        for (int a = 0; a < n; a++) {

            if (a > 0 && nums[a] == nums[a - 1]) {
                continue;
            }

            for (int b = a + 1; b < n; b++) {

                if (b > a + 1 && nums[b] == nums[b - 1]) {
                    continue;
                }

                int sum = target - nums[a] - nums[b];

                int d = n - 1;
                for (int c = b + 1; c < n; c++) {

                    if (c > b + 1 && nums[c] == nums[c - 1]) {
                        continue;
                    }

                    while (d > c && nums[c] + nums[d] > sum) {
                        d--;
                    }

                    if (c == d) {
                        break;
                    }
                    if (nums[c] + nums[d] == sum) {
                        ans.add(Arrays.asList(nums[a], nums[b], nums[c], nums[d]));
                    }

                }
            }
        }

        return ans;
    }

}

File: javacode/src/main/java/com/xiong/kotlin/leet/Leet18_fourSum.kt
package com.xiong.kotlin.leet

import java.util.*
import kotlin.test.currentStackTrace

class Leet18_fourSum {

    fun fourSum(nums: IntArray, target: Int): List<List<Int>> {

        val ans = arrayListOf<List<Int>>()

        val n = nums.size

        Arrays.sort(nums)

        for ( a in 0 until n - 3){

            if (a > 0 && nums[a] == nums[a - 1]) {
                continue
            }

            for (b in a + 1 until n - 2) {

                if (b > a + 1 && nums[b] == nums[b - 1]) {
                    continue
                }

                val sum = target - nums[a] - nums[b]
                var d = n - 1
                for (c in b + 1 until n - 1){

                    if (c > b + 1 && nums[c] == nums[c - 1]) {
                        continue
                    }

                    while (d > c && nums[c] + nums[d] > sum) {
                        d--
                    }
                    if (d == c) {
                        break
                    }

                    if (nums[c] + nums[d] == sum) {
                        ans.add(arrayListOf(nums[a], nums[b], nums[c], nums[d]))
                    }

                }
            }

        }

        return ans

    }
}